src/AccessControl/tests/test_formatter.py

```diff
@@ -36,14 +36,46 @@ def test_positional_argument_regression(self):
             'bar foo'
         )
 
-    def test_prevents_bad_string_formatting(self):
+    def test_prevents_bad_string_formatting_attribute(self):
         from AccessControl.safe_formatter import SafeFormatter
         self.assertRaises(Unauthorized,
                           SafeFormatter('{0.__class__}').safe_format,
                           'foo')
 
-    def test_prevents_bad_unicode_formatting(self):
+    def test_prevents_bad_unicode_formatting_attribute(self):
         from AccessControl.safe_formatter import SafeFormatter
         self.assertRaises(Unauthorized,
                           SafeFormatter(u'{0.__class__}').safe_format,
                           u'foo')
+
+    def test_prevents_bad_string_formatting_item(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        from persistent.mapping import PersistentMapping
+        foo = PersistentMapping({'__private__': 'keep off'})
+        self.assertRaises(Unauthorized,
+                          SafeFormatter('{0[__private__]}').safe_format,
+                          foo)
+
+    def test_prevents_bad_unicode_formatting_item(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        from persistent.mapping import PersistentMapping
+        foo = PersistentMapping({'__private__': 'keep off'})
+        self.assertRaises(Unauthorized,
+                          SafeFormatter(u'{0[__private__]}').safe_format,
+                          foo)
+
+    def test_prevents_bad_string_formatting_key(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        from persistent.list import PersistentList
+        foo = PersistentList(['__private__'])
+        self.assertRaises(Unauthorized,
+                          SafeFormatter('{0[0]}').safe_format,
+                          foo)
```

Comment:
Why would we want to protect this string list item?  AFAIK, "values" (rather than names / keys) don't get such protection under our security model.

Reply:
Say you get a list of items from a folderish content type, with `contentValues()` or maybe a custom method. If you use that list in a Zope2 page template or a Python skin Script, it seems good that AccessControl checks if you are allowed to access an item from that list.

Ah, I started wondering what happens if I try to circumvent this patch by simply getting a list of items, and asking for a private item by index, without bothering to use `str.format`. I see that, also without this patch, checks are done.  I tried this code in a Script Python in a Plone Site:

```
print context.contentValues()[index]
return printed
```

When I execute this script as anonymous user and pass an index that gives me a public page, I can see it just fine. (Well, `<Document at public>` here.) When instead I pass an index that would give me a private page, I get an Unauthorized. When I try that as admin user, I see it fine again.

I can partially get around that by doing a slice, because `guarded_getitem` does not check them: `context.contentValues()[4:]` will give me something like `[<Document at /Plone/private>]`. But that is still only a list, so when I want to actually do something with it, like accessing the Title, I need to get the item, which will properly call `guarded_getitem` again, which will block me.

So it looks like values are checked after all, at least in this example case.

Note that `guarded_getitem` does no checks when both the container and the contained items are simple Python objects, which is why I had to test this with a `PersistentList` instead of a standard Python list.